nemo-physical/src/dictionary/hash_map_dictionary.rs

```diff
@@ -0,0 +1,413 @@
+
+/// Address size of pages in the string buffer
+const PAGE_ADDR_BITS: usize = 25; // 32MB
+/// Size of pages in the string buffer
+const PAGE_SIZE: usize = 1 << PAGE_ADDR_BITS;
+/// Bit mask that keeps only the (lower) PAGE_ADDR_BITS-1 bits, for extracting a string's length
+const LENGTH_BITS_MASK: u64 = (1 << (PAGE_ADDR_BITS - 1)) - 1;
+
+/// A buffer for string data using compact memory regions that are managed in pages.
+/// New buffers need to be initialized, upn which they will receive an identifying buffer id
+/// that is used whenever the data accessed.
+///
+/// The implementaion is not fully thread-safe, but it is thread-safe as long as each buffer
+/// is used in only one thread. That is, parallel threads can safely create buffers (which will
+/// have different ids), as long as all their operations use the buffer id that they were given.
+struct StringBuffer {
+    /// Vector of buffer ids and string buffers
+    pages: Vec<(usize, String)>,
+    /// Single temporary string per buffer. [StringRef] uses this for representing strings that are not in the buffer.
+    ///
+    /// TODO: It would be possible and more elegant to have a special alternative key implementation for our hashmap,
+    /// where the key has the relevant data instead of pointing to a temporary buffer.
+    tmp_strings: Vec<String>,
+    /// Currently active page for each buffer
+    cur_pages: Vec<usize>,
+    /// Lock to guard page assignment operations when using multiple threads
+    lock: AtomicBool,
+}
+impl StringBuffer {
+    /// Constructor.
+    const fn new() -> Self {
+        StringBuffer {
+            pages: Vec::new(),
+            tmp_strings: Vec::new(),
+            cur_pages: Vec::new(),
+            lock: AtomicBool::new(false),
+        }
+    }
+
+    /// Initializes a new buffer and returns a handle that can henceforth be used to access it.
+    fn init_buffer(&mut self) -> usize {
+        self.acquire_page_lock();
+        let buf_id = self.cur_pages.len();
+        self.pages.push((buf_id, String::with_capacity(PAGE_SIZE)));
+        self.cur_pages.push(self.pages.len() - 1);
+        self.tmp_strings.push(String::new());
+        self.release_page_lock();
+        buf_id
+    }
+
+    /// Frees the memory used by the pages of the dropped buffer.
+    /// No other pages are affected or moved.
+    fn drop_buffer(&mut self, buffer: usize) {
+        self.acquire_page_lock();
+        for (b, s) in self.pages.iter_mut() {
+            if buffer == *b {
+                s.clear();
+                s.shrink_to_fit();
+                *b = usize::MAX;
+            }
+        }
+        self.release_page_lock();
+    }
+
+    /// Inserts a string into the buffer and returns a [StringRef] that points to it.
+    ///
+    /// TODO: Allocation of new pages could re-use freed pages instead of always appending.
+    fn push_str(&mut self, buffer: usize, s: &str) -> StringRef {
+        let len = s.len();
+        assert!(len < 1 << 25);
```

Comment:
I assume this should explicitly reference `PAGE_SIZE`, since changing that would potentially break the `assert!`.